cpp/include/cuvs/neighbors/ivf_list.hpp

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/core/serialize.hpp>
+#include <raft/util/integer_utils.hpp>
+
+#include <thrust/fill.h>
+
+#include <atomic>
+#include <fstream>
+#include <limits>
+#include <memory>
+#include <type_traits>
+
+namespace cuvs::neighbors::ivf {
+
+/**
+ * Default value filled in the `indices` array.
+ * One may encounter it trying to access a record within a list that is outside of the
+ * `size` bound or whenever the list is allocated but not filled-in yet.
+ */
+template <typename IdxT>
+constexpr static IdxT kInvalidRecord =
+  (std::is_signed_v<IdxT> ? IdxT{0} : std::numeric_limits<IdxT>::max()) - 1;
+
+/** The data for a single IVF list. */
+template <template <typename, typename...> typename SpecT,
+          typename SizeT,
+          typename... SpecExtraArgs>
+struct list {
```

Comment:
Need to define the contract interface in include/ and push all impls to source files. This is also what improves compile times as the source files get compiled once and any and all users just rely on (and adhere to) the definitions in `include/`.